KAFKA-16362: Fix type-unsafety in KStreamKStreamJoin caused by isLeftSide

raminqaf opened the pull request:
#### Description
The introduced changes provide a cleaner definition of the join side in KStreamKStreamJoin. Before, this was done by using a Boolean flag, which led to returning a raw LeftOrRightValue without generic arguments because the generic type arguments depended on the boolean input.

Tests of @gharris1727 from his [PR](https://github.com/apache/kafka/pull/15513) have been adopted to check the type-sefty of the joins.

#### Summary of testing strategy
Unit tests have been added for the new _JoinSide_ enum.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

wernerdv committed:
separate left and right stream classes

raminqaf commented on streams/src/main/java/org/apache/kafka/streams/state/internals/LeftOrRightValue.java:
Retroperspectivly, we can refactor this class and break it into two subclasses (`LeftJoinValue<LV, Object>` & `RightJoinValue<Object,RV>`) that implement the same interface (`JoinValue<LV,RV>`). We can use the interface in the enum to make the objects.
I already created these on my machine, but I wanted to avoid making significant changes in a single PR and moving too many classes simultaneously.

gharris1727 replied:
I assumed that this would be necessary. I think either could work, and it's up to the your taste. If this enum is only useful for this one class, that could indicate that the the enum and joiner are too coupled. I'm not familiar with Streams, but maybe there's another class that could benefit from this functionality?

raminqaf replied:
@gharris1727 Thanks for the fast review! I have addressed your comments. In my opinion, the bigger refactoring should be done in a follow-up ticket/issue, and this one should specifically solve the casting issue.
This is also my first contribution here, so I don't want to move a lot of parts and have a forever open PR.

gharris1727 replied:
Sure, I don't think that we have to apply this change to other classes or look for more call-sites. I was just thinking out loud about the benefits of trying to make the fix reusable.

Even if we're only changing KStreamKStreamJoin, I would much rather merge a large diff, with a type-safe and simple implementation than a small diff that is type-unsafe and complicated. I love diff-golfing for bugfixes, but I don't think it's appropriate for refactoring.

gharris1727 commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
```diff
@@ -250,7 +252,7 @@ private void emitNonJoinedOuterRecords(
 
                     final VOut nullJoinedValue;
-                    if (isLeftSide) {
+                    if (joinSide.isLeftSide()) {
```
This is another unsafe place that should be part of the enum.
Try and eliminate all of the `unchecked` warnings in this class.

gharris1727 commented on streams/src/main/java/org/apache/kafka/streams/state/internals/LeftOrRightValue.java:
```diff
@@ -39,45 +39,6 @@ private LeftOrRightValue(final V1 leftValue, final V2 rightValue) {
-     * @param <V1>      the type of the value
-     * @return a new {@link LeftOrRightValue} instance
-     */
-    public static <V1, V2> LeftOrRightValue<V1, V2> makeLeftValue(final V1 leftValue) {
```
I think you should keep the `makeLeftValue` and `makeRightValue` functions as entrypoints to this class. Only the the impossible-to-type `make` function should be removed.

gharris1727 commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
```diff
@@ -155,7 +156,8 @@ public void process(final Record<K, V1> record) {
                         // range over all values of the same key even after failure, since the other window-store
                         // is only cleaned up by stream time, so this is okay for at-least-once.
-                        store.putIfAbsent(TimestampedKeyAndJoinSide.make(!isLeftSide, record.key(), otherRecordTimestamp), null);
+                        final JoinSide otherJoinSide = joinSide.isLeftSide() ? JoinSide.RIGHT : JoinSide.LEFT;
```
This could be a `JoinSide#complement` or `JoinSide#opposite` function, in case there are other situations where we need to know what the other side of the join is doing.

gharris1727 commented on streams/src/main/java/org/apache/kafka/streams/state/internals/JoinSide.java:
```diff
@@ -0,0 +1,79 @@
+     *
+     * @return true if this JoinSide represents the left side, otherwise false
+     */
+    public boolean isLeftSide() {
```
This is a very permissive "escape hatch" from this enum, and could be hiding other abstractions that could be pulled into the enum.

raminqaf replied:
Closed the "hatch" forever! 😄 I had it removed!

gharris1727 replied:
It looks like you inlined the `== LEFT` check, which doesn't really fix the problem. There really shouldn't be any `if (left)` statements in the caller at all.

gharris1727 commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImplJoin.java:
```diff
@@ -173,7 +174,7 @@ public <K, V1, V2, VOut> KStream<K, VOut> join(final KStream<K, V1> lhs,
         final JoinWindowsInternal internalWindows = new JoinWindowsInternal(windows);
         final KStreamKStreamJoin<K, V1, V2, VOut> joinThis = new KStreamKStreamJoin<>(
-            true,
+            JoinSide.LEFT,
```
I think these call-sites are much more clear with the explicit "left" and "right". It's interesting that the variables here are joinThis and joinOther, instead of joinLeft and joinRight.

gharris1727 replied:
Sorry, I should have clarified that I wasn't necessarily asking you to change this, I was just making an observation.

The ProcessorParameters and StreamStreamJoinNodeBuilder use "this" and "other", and I don't know whether it makes sense to change the naming there either.

gharris1727 reviewed (commented):
Hi @raminqaf Thanks for picking up this ticket.

I'm not very familiar with Streams, so my comments aren't blocking, and I'll let a Streams committer take a look at this before merge.

Cool idea to use an enum for the refactor!

gharris1727 commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImplJoin.java:
```diff
@@ -41,6 +41,7 @@
 import java.util.Optional;
 import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
+import org.apache.kafka.streams.state.internals.JoinSide;
```
I just noticed that the JoinSide package is `state`, while this class is `kstream`. Since this enum is so tightly bound to this class, it should probably be in the same package.

gharris1727 commented on streams/src/main/java/org/apache/kafka/streams/state/internals/JoinSide.java:
```diff
@@ -0,0 +1,79 @@
+ * An enum representing the side of a join operation.
+ * It provides methods to create instances of {@link LeftOrRightValue} based on the side specified.
+ */
+@SuppressWarnings("unchecked")
```
It looks like this is just moving the unchecked warning around, rather than fixing it.
`<V, V1, V2> LeftOrRightValue<V1, V2> make(final V leftValue);` is impossible to type in a similar way to `<V> LeftOrRightValue make(final boolean isLeftSide, final V value)`, because it will always require casting V to V1 or V2.

It might not be possible to do this with Enum, as it can't take type arguments.

gharris1727 commented on streams/src/main/java/org/apache/kafka/streams/state/internals/LeftOrRightValue.java:
```diff
@@ -89,8 +74,8 @@ public V2 getRightValue() {
         return "<"
-            + ((leftValue != null) ? "left," + leftValue : "right," + rightValue)
-            + ">";
+                + ((leftValue != null) ? JoinSide.LEFT + "," + leftValue : JoinSide.RIGHT + "," + rightValue)
```
This doesn't seem necessary. LeftOrRightValue can be completely unaware of the JoinSide enum.

gharris1727 commented on streams/src/main/java/org/apache/kafka/streams/state/internals/LeftOrRightValueDeserializer.java:
```diff
@@ -62,13 +62,13 @@ public LeftOrRightValue<V1, V2> deserialize(final String topic, final byte[] dat
         }
 
         return (data[0] == 1)
-            ? LeftOrRightValue.makeLeftValue(leftDeserializer.deserialize(topic, rawValue(data)))
```
Same comment as in LeftOrRightValue, I don't think this is necessary.

gharris1727 commented:
This PR looks like it'll have merge conflicts with #15510. Since that is fixing a bug it should probably have higher priority than this refactor, can you rebase on top of their changes? Alternatively we can wait for that to land first before returning to this PR.

raminqaf commented:
@gharris1727 I have broken down the KStreamKstreamJoin class into two classes. For now, I just moved the code (+the fix in #15510) to see if all the tests pass and if I am going in the correct direction. I managed to get rid of the unsafe type casts appropriately. 
I created [two outer join stores](https://github.com/raminqaf/kafka/blob/dc608850c5e7af1a6f589b30f9f7f0921f64942c/streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java#L76-L79) (one for the left side and one for the right side). This will remove one of the unsafe casts. The other unsafe cast on the right side of the join [was removed based on the join's left or right value](https://github.com/raminqaf/kafka/blob/dc608850c5e7af1a6f589b30f9f7f0921f64942c/streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamRightJoin.java#L215-L218).

I changed the PR into a draft because it depends on #15510. Please have a look and let me know if this is going in the correct direction.

raminqaf commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoin.java:
```diff
@@ -0,0 +1,201 @@
+            }
+
+            boolean needOuterJoin = outer;
+            try (final WindowStoreIterator<V2> iter = otherWindowStore.fetch(leftRecord.key(), timeFrom, timeTo)) {
```
I also have a couple of questions. When is this store getting filled? I couldn't find anywhere where we call `.put()` for this window store.

mcmmining replied:
> I also have a couple of questions. When is this store getting filled? I couldn't find anywhere where we call `.put()` for this window store.

raminqaf commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
```diff
@@ -16,276 +16,98 @@
+        protected long getOuterJoinLookBackTimeMs(final TimestampedKeyAndJoinSide<K> timestampedKeyAndJoinSide) {
             // depending on the JoinSide we fill in the outerJoinLookBackTimeMs
-            if (timestampedKeyAndJoinSide.isLeftSide()) {
+            if (timestampedKeyAndJoinSide.getJoinSide() == JoinSide.LEFT) {
```
A side note: This decision could not be removed. In the KStreamKStreamRightJoin, there are some tests that fetch the value from the rightOuterStore, and the value contains the left-side value.

In the KStreamKStreamLeftJoin this call was not needed.

raminqaf commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamRightJoin.java:
```diff
@@ -0,0 +1,221 @@
+                    final long outerJoinLookBackTimeMs = getOuterJoinLookBackTimeMs(timestampedKeyAndJoinSide);
+                    if (sharedTimeTracker.minTime + outerJoinLookBackTimeMs + joinGraceMs
+                            >= sharedTimeTracker.streamTime) {
+                        if (timestampedKeyAndJoinSide.getJoinSide() == JoinSide.LEFT) {
```
A side note: This decision should ideally be removed. The reason is still there is because I am initialising the left- and right outer stores with the same name. See the 'init()' method in the parent class.

In the KStreamKStreamLeftJoin this call was not needed.

gharris1727 commented on streams/src/main/java/org/apache/kafka/streams/state/internals/TimestampedKeyAndJoinSide.java:
```diff
@@ -33,28 +34,36 @@
      */
-    public static <K> TimestampedKeyAndJoinSide<K> make(final boolean leftSide, final K key, final long timestamp) {
-        return new TimestampedKeyAndJoinSide<>(leftSide, key, timestamp);
+    public static <K> TimestampedKeyAndJoinSide<K> make(final JoinSide joinSide, final K key, final long timestamp) {
```
Since this is only used in tests now, I think you can remove this and replace the test call-sites with the new functions.

Make sure to copy the javadoc to the new signatures too.

gharris1727 commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
```diff
@@ -16,276 +16,98 @@
+    protected abstract class KStreamKStreamJoinProcessor extends ContextualProcessor<K, V1, K, VOut> {
+        protected WindowStore<K, V2> otherWindowStore;
+        protected Sensor droppedRecordsSensor;
+        protected Optional<KeyValueStore<TimestampedKeyAndJoinSide<K>, LeftOrRightValue<V1, V2>>> leftOuterJoinStore =
+                Optional.empty();
+        protected Optional<KeyValueStore<TimestampedKeyAndJoinSide<K>, LeftOrRightValue<V2, V1>>> rightOuterJoinStore =
+                Optional.empty();
```
The left store is only used by the Left class, and the right store is only used by the Right class. These can be combined into a single outerJoinStore, and the KStreamKStreamJoinProcessor can take generic arguments <Va, Vb> that determine the types in the joinStore signature.

The LeftJoinProcessor can then set Va=V1 and Vb=V2, while the RightJoinProcessor sets Va=V2 and Vb=V1.

gharris1727 commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoin.java:
```diff
@@ -0,0 +1,201 @@
+                        continue; // there are possibly candidates left on the other outerJoin-side
+                    }
+
+                    final VOut nullJoinedValue = joiner.apply(key, value.getLeftValue(), value.getRightValue());
```
Pull this out into getNullJoinedValue like the other side.

gharris1727 commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoin.java:
```diff
@@ -0,0 +1,201 @@
+            final boolean outer,
+            final Optional<String> outerJoinWindowName,
+            final TimeTrackerSupplier sharedTimeTrackerSupplier) {
+        super(otherWindowName, windows, windows.beforeMs, windows.afterMs, joiner, outerJoinWindowName,
```
I like this constructor refactor for the windows.beforeMs and the windows.afterMs! :+1+

gharris1727 reviewed (commented):
Hi @raminqaf I think this is in the right direction. I have some high-level concerns that I'll mention here instead of pointing out individual problems:

Can you please revert changes where things are re-ordered or have spacing changes, but don't have any semantic changes? I saw this for imports, constructor arguments, and some multi-line statements. It would be easier to review and generate less merge conflicts if these lines did not change.

I think the current refactor of the process and emitNonJoinedOuterRecords methods introduces too much duplication. I think you can keep most of the implementation inside of KStreamKStreamJoinProcessor if you give it generic arguments, and have small abstract methods whenever the generics are insufficient. I pointed out two simplifications for the outerJoinStore and getNullJoinedValue that I think demonstrate the idea, but there are certainly other abstract methods needed.

Commit:
Merge branch 'trunk' of https://github.com/apache/kafka into refactor/KAFKA-16362

# Conflicts:
#	streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java

Commit:
Move init method to abstract class

Commit:
Move class fields to super class

Commit:
Move more code to super class

Commit:
Move inner join logic to super class

Commit:
Clean up

Commit:
Clean up

Commit:
Clean up

raminqaf commented on streams/src/main/java/org/apache/kafka/streams/state/internals/TimestampedKeyAndJoinSide.java:
```diff
@@ -52,7 +52,26 @@ private TimestampedKeyAndJoinSide(final boolean leftSide, final K key, final lon
         return new TimestampedKeyAndJoinSide<>(leftSide, key, timestamp);
     }
-
+    /**
+     * Create a new {@link TimestampedKeyAndJoinSide} instance for the left join side if the provide {@code key} is not {@code null}.
+     *
+     * @param key      the key
+     * @param <K>      the type of the key
+     * @return a new {@link TimestampedKeyAndJoinSide} instance for the left join side if the provide {@code key} is not {@code null}
+     */
+    public static <K> TimestampedKeyAndJoinSide<K> makeLeft(final K key, final long timestamp) {
+        return new TimestampedKeyAndJoinSide<>(true, key, timestamp);
+    }
+    /**
+     * Create a new {@link TimestampedKeyAndJoinSide} instance for the right join side if the provide {@code key} is not {@code null}.
+     *
+     * @param key      the key
+     * @param <K>      the type of the key
+     * @return a new {@link TimestampedKeyAndJoinSide} instance for the right join side if the provide {@code key} is not {@code null}
+     */
+    public static <K> TimestampedKeyAndJoinSide<K> makeRight(final K key, final long timestamp) {
+        return new TimestampedKeyAndJoinSide<>(false, key, timestamp);
+    }
```
@gharris1727 I have added these static factory methods for a cleaner definition. If you prefer, I can have them removed and use the `make` method.

raminqaf commented on streams/src/main/java/org/apache/kafka/streams/state/internals/LeftOrRightValue.java:
```diff
@@ -63,21 +63,6 @@ public static <V1, V2> LeftOrRightValue<V1, V2> makeRightValue(final V2 rightVal
         return new LeftOrRightValue<>(null, rightValue);
     }
 
-    /**
-     * Create a new {@link LeftOrRightValue} instance with the V value as {@code leftValue} if
-     * {@code isLeftSide} is True; otherwise {@code rightValue} if {@code isLeftSide} is False.
-     *
-     * @param value the V value (either V1 or V2 type)
-     * @param <V>   the type of the value
-     * @return a new {@link LeftOrRightValue} instance
-     */
-    public static <V> LeftOrRightValue make(final boolean isLeftSide, final V value) {
-        Objects.requireNonNull(value, "value is null");
-        return isLeftSide
-            ? LeftOrRightValue.makeLeftValue(value)
-            : LeftOrRightValue.makeRightValue(value);
-    }
-
```
Removed as promised 😄

raminqaf commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoin.java:
A clean abstraction. Each join side defines its "this side value" and "other side value".

gharris1727 replied:
The left and right classes are perfect in my opinion.

raminqaf commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
No more un safe type casting to be found in this class. Not that much has changed in this abstract class. I have moved some code in methods. 

Side Note: I don't have the correct formatter configured :/ I couldn't find any contribution notes to set the correct one.

gharris1727 replied:
> No more un safe type casting to be found in this class. Not that much has changed in this abstract class. I have moved some code in methods.

Very nice! I really like the new types, and the abstract methods are very minimal.

> Side Note: I don't have the correct formatter configured :/ I couldn't find any contribution notes to set the correct one.

I'm not aware of a standard formatter either. I remember tweaking the IntelliJ default rules slightly, but taking a look at them now I'm not seeing evidence of what I changed.

Commit:
Clean up

Commit:
Formatting

gharris1727 commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
```diff
@@ -275,24 +226,47 @@ private void emitNonJoinedOuterRecords(
-                        (V1) leftOrRightValue.getRightValue(),
-                        (V2) leftOrRightValue.getLeftValue());
+
+        private void performInnerJoin(final Record<K, VThis> thisRecord, final long thisRecordTimestamp) {
```
nit: I think this a bit of a confusing method name, since it mentions InnerJoin, but it has needOuterJoin, and interacts with the outerJoinStore. I think this should be inlined back into process().

gharris1727 commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
```diff
@@ -240,27 +185,33 @@ private void emitNonJoinedOuterRecords(
                     final K key = timestampedKeyAndJoinSide.getKey();
-                    final LeftOrRightValue<V1, V2> leftOrRightValue = next.value;
-                    final VOut nullJoinedValue = getNullJoinedValue(key, leftOrRightValue);
+                    final LeftOrRightValue<VL, VR> leftOrRightValue = next.value;
+                    final VThis thisValue = getThisValue(leftOrRightValue);
+                    final VOther otherValue = getOtherValue(leftOrRightValue);
+                    final VOut nullJoinedValue = joiner.apply(key, thisValue, otherValue);
                     context().forward(
-                        record.withKey(key).withValue(nullJoinedValue).withTimestamp(timestamp)
+                            record.withKey(key).withValue(nullJoinedValue).withTimestamp(timestamp)
                     );
```
nit: This could be brought out to a method which depends only on `record` and `next`.

gharris1727 commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
```diff
@@ -43,161 +42,106 @@
-
-            if (outer && record.key() == null && record.value() != null) {
-                context().forward(record.withValue(joiner.apply(record.key(), record.value(), null)));
+        public void process(final Record<K, VThis> thisRecord) {
```
nit: Personally I don't think calling it thisRecord and thisRecordTimestamp are so necessary, because the type checker will make sure we're using the right one.

gharris1727 reviewed (commented):
Hey @raminqaf this is looking really good! I like how there is a separation between the this/other used in the joiner and Processor signatures, and left/right used in the outerJoinStore.

I really only had nits, as this accomplishes the type-safety I was looking for.

I do think there are still a lot of non-semantic changes to ordering and indenting that I think could be addressed before a Streams committer takes a look at this, and would make their review faster and easier.

Commit:
Merge branch 'trunk' of https://github.com/apache/kafka into refactor/KAFKA-16362

Commit:
Address reviews and refactor code into methods

gharris1727 commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
```diff
@@ -43,95 +42,73 @@
+    KStreamKStreamJoin(final String otherWindowName, final TimeTrackerSupplier sharedTimeTrackerSupplier,
+                       final Optional<String> outerJoinWindowName, final long joinBeforeMs,
+                       final long joinAfterMs, final JoinWindowsInternal windows, final boolean outer,
+                       final ValueJoinerWithKey<? super K, ? super VThis, ? super VOther, ? extends VOut> joiner) {
```
Revert the re-ordering here that isn't necessary, so it's easy to see the signature change for `joiner` and the removal of `isLeftSide`.

We should also put each constructor argument on it's own to follow the previous style.

gharris1727 commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
```diff
@@ -43,95 +42,73 @@
+        this.windowsBeforeMs = windows.beforeMs;
+        this.windowsAfterMs = windows.afterMs;
+        this.outer = outer;
+        this.joiner = joiner;
```
nit: Reorder this to make the diff smaller.

gharris1727 commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
```diff
@@ -43,95 +42,73 @@
         private TimeTracker sharedTimeTracker;
+        private Optional<KeyValueStore<TimestampedKeyAndJoinSide<K>, LeftOrRightValue<VL, VR>>> outerJoinStore =
+                Optional.empty();
+        private WindowStore<K, VOther> otherWindowStore;
```
nit: reorder this to make the diff smaller.

gharris1727 commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
```diff
@@ -184,20 +146,34 @@ public void process(final Record<K, V1> record) {
                     if (!outerJoinStore.isPresent() || timeTo < sharedTimeTracker.streamTime) {
-                        context().forward(record.withValue(joiner.apply(record.key(), record.value(), null)));
+                        context().forward(
+                                record.withValue(joiner.apply(record.key(), record.value(), null)));
```
nit: revert

gharris1727 commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
```diff
@@ -43,95 +42,73 @@
+                                EMIT_INTERVAL_MS_KSTREAMS_OUTER_JOIN_SPURIOUS_RESULTS_FIX,
+                                1000L
+                        )
                 );
```
nit: revert changes here

gharris1727 commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
```diff
@@ -184,20 +146,34 @@ public void process(final Record<K, V1> record) {
+        @Override
+        public void close() {
+            sharedTimeTrackerSupplier.remove(context().taskId());
+        }
```
nit: move this down near the bottom for a smaller diff

gharris1727 commented on streams/src/main/java/org/apache/kafka/streams/state/internals/LeftOrRightValue.java:
```diff
@@ -89,8 +74,8 @@ public V2 getRightValue() {
-            + ((leftValue != null) ? "left," + leftValue : "right," + rightValue)
-            + ">";
+                + ((leftValue != null) ? "left," + leftValue : "right," + rightValue)
+                + ">";
```
nit: revert this indent change

gharris1727 commented on streams/src/main/java/org/apache/kafka/streams/state/internals/LeftOrRightValue.java:
```diff
@@ -103,7 +88,7 @@ public boolean equals(final Object o) {
         final LeftOrRightValue<?, ?> that = (LeftOrRightValue<?, ?>) o;
         return Objects.equals(leftValue, that.leftValue) &&
-            Objects.equals(rightValue, that.rightValue);
+                Objects.equals(rightValue, that.rightValue);
```
nit: revert this indent change

gharris1727 reviewed (commented):
Here are most of the places where I see non-semantic changes that can be eliminated from the diff.

But please look through the whole diff yourself: ask if each line is a necessary part of the refactor, or if it's something that got moved but could be left as-is.

Commit:
Merge branch 'trunk' of https://github.com/apache/kafka into refactor/KAFKA-16362

Commit:
Indent code for smaller diff

Commit:
Update files

Commit:
Update files

Commit:
fix super

Commit:
Update files

Commit:
fix indent

raminqaf commented:
@gharris1727 Thanks for the feedback! I reverted all the changes you requested and reverted a couple of other indentation problems that caused a diff. I can even go further and revert & inline the introduce private methods (i.e., `emitInnerJoin` or `putInOuterJoinStore`)

gharris1727 commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
```diff
@@ -187,17 +157,25 @@ public void process(final Record<K, V1> record) {
-                        outerJoinStore.ifPresent(store -> store.put(
-                            TimestampedKeyAndJoinSide.make(isLeftSide, record.key(), inputRecordTimestamp),
-                            LeftOrRightValue.make(isLeftSide, record.value())));
+                        putInOuterJoinStore(record, inputRecordTimestamp);
```
This can just take `record` and call `record.timestamp()` itself.

gharris1727 commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
```diff
@@ -275,22 +240,34 @@ private void emitNonJoinedOuterRecords(
+                    >= sharedTimeTracker.streamTime;
         }
 
+
```
nit: extra line

gharris1727 commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
```diff
@@ -221,41 +199,28 @@ private void emitNonJoinedOuterRecords(
+                    final KeyValue<TimestampedKeyAndJoinSide<K>, LeftOrRightValue<VL, VR>> nextKeyValue = it.next();
+                    final TimestampedKeyAndJoinSide<K> timestampedKeyAndJoinSide = nextKeyValue.key;
+                    sharedTimeTracker.minTime = timestampedKeyAndJoinSide.getTimestamp();
+                    if (isOuterJoinWindowOpenForSide(timestampedKeyAndJoinSide, true) && isOuterJoinWindowOpenForSide(timestampedKeyAndJoinSide, false)) {
```
Can this condition ever fire?
If timestampedKeyAndJoinSide.isLeftSide() is true, then only the first condition can be true.
If timestampedkeyAndJoinSide.isLeftSide() is false, then only the second condition can be true.

These were state variables shared across multiple iterations and incorporated multiple timestampedKeyAndJoinSide before, now it's a function of just a single timestampedKeyAndJoinSide. Without knowing the context here, I would guess that would change the behavior.

raminqaf replied:
Yeah, you are right! That's a good catch. I was depending too much on the tests. I couldn't find any test that triggered the condition even after reverting the changes.

gharris1727 commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
```diff
@@ -144,26 +129,11 @@ public void process(final Record<K, V1> record) {
+            final long timeFrom = Math.max(0L, inputRecordTimestamp - joinBeforeMs);
+            final long timeTo = Math.max(0L, inputRecordTimestamp + joinAfterMs);
+            try (final WindowStoreIterator<VOther> iter = otherWindowStore.fetch(record.key(), timeFrom, timeTo)) {
+                final boolean needOuterJoin = outer && !iter.hasNext();
```
I think this makes sense to make this a final variable, good find!

gharris1727 reviewed (commented):
I think I'm reaching the end of my useful review comments, and this is nearly ready for a streams reviewer. @mjsax @ableegoldman @guozhangwang do any of you have time to take a look at this?

ableegoldman commented:
Thanks Greg. If no one's had time to look at this by next week, we'll assign a reviewer during the next Kafka Streams hangout

Commit:
Revert window open condition

cadonna commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
```diff
@@ -43,7 +42,7 @@
 import static org.apache.kafka.streams.processor.internals.metrics.TaskMetrics.droppedRecordsSensor;
 
-class KStreamKStreamJoin<K, V1, V2, VOut> implements ProcessorSupplier<K, V1, K, VOut> {
+abstract class KStreamKStreamJoin<K, VL, VR, VOut, VThis, VOther> implements ProcessorSupplier<K, VThis, K, VOut> {
```
I do not get the difference between `VL` and `VThis` as well as `VR` and `VOther`. Could you please elaborate?

cadonna replied:
Also a better name for `VL` and `VR` would be `VLeft` and `VRight`, respectively. `VR` is sometimes used for result value type.

raminqaf replied:
We needed a way to detect the type of the side being called dynamically so that the super class can handle them based on the caller. This also ensures type safety.

When this class is called by the _left side_ of the join `VThis` would be _left_ left type (`VLeft`) and `VOther` would be _right_ type (`VRight`).

```java
class KStreamKStreamLeftJoinSide<K, VLeft, VRight, VOut> extends KStreamKStreamJoin<K, VLeft, VRight, VOut, VLeft, VRight> {
																											--^--  --^---
																											VThis VOther
```

Whereas for the _right side_ of the join `VThis` would be _right_ left type (`VRight`) and `VOther` would be _left_ type (`VLeft`).

```java
class KStreamKStreamRightJoinSide<K, VLeft, VRight, VOut> extends KStreamKStreamJoin<K, VLeft, VRight, VOut, VRight, VLeft>
																											--^--  --^---
																											VThis VOther
```

raminqaf replied:
It is more tangible if we look at the abstract methods:
```java
protected abstract VThis getThisValue(final LeftOrRightValue<? extends VLeft, ? extends VRight> leftOrRightValue);

protected abstract VOther getOtherValue(final LeftOrRightValue<? extends VLeft, ? extends VRight> leftOrRightValue);
```
Depending on which join side implements these methods, `VThis` and `VOther` can change. For instance, the `getThisValue()` method on the _left side_ of the join would be:
```java
 @Override
public VLeft getThisValue(final LeftOrRightValue<? extends VLeft, ? extends VRight> leftOrRightValue) {
      return leftOrRightValue.getLeftValue();
}
```
Respectively, on the _right side_ we have
```java
  @Override
  public VRight getThisValue(final LeftOrRightValue<? extends VLeft, ? extends VRight> leftOrRightValue) {
      return leftOrRightValue.getRightValue();
  }
  ````

gharris1727 replied:
Just to throw my own explanation in here:
VLeft and VRight are "absolute" in that they are the left and right types of the overall join. Both sides of the join have equivalent VLeft and VRight types, because they share a common outerJoinStore instance.

VThis and VOther are "relative" in that they are the type of records entering "this" side of the join, and the "other" side of the join, and this is necessarily swapped for the other side of the join.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoin.java:
```diff
@@ -0,0 +1,72 @@
+
+import java.util.Optional;
+
+class KStreamKStreamLeftJoin<K, VL, VR, VOut> extends KStreamKStreamJoin<K, VL, VR, VOut, VL, VR> {
```
Do you mean that this is the left side of the join or that this is a left join? That is a bit confusing. Maybe a better name might be `KStreamKStreamJoinLeftSide`.

raminqaf replied:
Good catch. What do you think about `KStreamKStreamLeftJoinSide` instead?

cadonna replied:
I would still prefer `KStreamKStreamJoinLeftSide`. It is the left side of a stream-stream join. The processor are also named `KStreamKStreamJoinLeftProcessor`.

cadonna reviewed (commented):
@raminqaf Thanks for the PR!

I did a quick pass. I will go more into details in the coming days.

Commit:
Merge branch 'trunk' of https://github.com/apache/kafka into refactor/KAFKA-16362

Commit:
Address reviews

cadonna commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
```diff
@@ -187,17 +157,25 @@ public void process(final Record<K, V1> record) {
-            final KeyValueStore<TimestampedKeyAndJoinSide<K>, LeftOrRightValue<V1, V2>> store,
-            final Record<K, V1> record) {
+            final KeyValueStore<TimestampedKeyAndJoinSide<K>, LeftOrRightValue<VLeft, VRight>> store,
+            final Record<K, ?> record) {
```
nit:
```java
private void emitNonJoinedOuterRecords(final KeyValueStore<TimestampedKeyAndJoinSide<K>, LeftOrRightValue<VLeft, VRight>> store, 
                                       final Record<K, ?> record) {

```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
```diff
@@ -187,17 +157,25 @@ public void process(final Record<K, V1> record) {
-            final KeyValueStore<TimestampedKeyAndJoinSide<K>, LeftOrRightValue<V1, V2>> store,
-            final Record<K, V1> record) {
+            final KeyValueStore<TimestampedKeyAndJoinSide<K>, LeftOrRightValue<VLeft, VRight>> store,
+            final Record<K, ?> record) {
```
Shouldn't that be
```java
final Record<K, VThis> record
```
?

raminqaf replied:
Good catch! Avoiding the wildcard here makes it easier to read and understand. I have changed it!

cadonna commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
```diff
@@ -275,20 +246,22 @@ private void emitNonJoinedOuterRecords(
-                        (V1) leftOrRightValue.getRightValue(),
-                        (V2) leftOrRightValue.getLeftValue());
-            }
+        private void forwardNonJoinedOuterRecords(final Record<K, ?> record, final KeyValue<? extends TimestampedKeyAndJoinSide<K>, ? extends LeftOrRightValue<VLeft, VRight>> nextKeyValue) {
```
I think it would be simpler to define this method as:
```java
private void forwardNonJoinedOuterRecords(final Record<K, VThis> record,
                                          final TimestampedKeyAndJoinSide<K> timestampedKeyAndJoinSide,
                                          final LeftOrRightValue<VLeft, VRight> leftOrRightValue) {
```
It makes the code a bit simpler and shorter.
Also here, why not `Record<K, VThis> record`. That is exactly the type used at the call site.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoinSide.java:
```diff
@@ -0,0 +1,72 @@
+
+    private class KStreamKStreamJoinLeftProcessor extends KStreamKStreamJoinProcessor {
+
+
```
```suggestion
```

cadonna reviewed (commented):
Thanks for the updates, @raminqaf!

Here my feedback!

Commit:
address reviews

Commit:
Merge branch 'trunk' of https://github.com/apache/kafka into refactor/KAFKA-16362

cadonna commented:
@raminqaf Could you please rebase this PR on current trunk to get the new build setup? We need to restart the builds because one was red.

Commit:
Merge branch 'trunk' of https://github.com/apache/kafka into refactor/KAFKA-16362

raminqaf commented:
> @raminqaf Could you please rebase this PR on current trunk to get the new build setup? We need to restart the builds because one was red.

Done!

cadonna reviewed (approved):
Thanks for the updates, @raminqaf !

LGTM!
